src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] values;
+    private int size;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        grow();
+        values[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index);
+        grow();
+        System.arraycopy(values, index,
+                values, index + 1,
+                size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] a = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            a[i] = list.get(i);
+        }
+        int varLength = a.length;
+        grow();
+        System.arraycopy(a,0,values,size, varLength);
+        size += varLength;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexWithSameSize(index);
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexWithSameSize(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexWithSameSize(index);
+        final T removedElement = values[index];
+        System.arraycopy(values, index + 1,
+                values, index,
+                values.length - index - 1);
+        values[values.length - 1] = null;
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (customEquals(element, values[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found in ArrayList");
+    }
+
+    private void grow() {
+        if (size == values.length) {
+            T[] temp = (T[]) new Object[values.length];
+            System.arraycopy(values, 0, temp, 0, size);
+            values = (T[]) new Object[values.length << 1];
+            System.arraycopy(temp, 0, values, 0, size);
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index: " + index);
+        }
+    }
+
+    private void checkIndexWithSameSize(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index: " + index);
+        }
+    }
```

Comment:
since you're using the checkIndex() method 
method once, you can use checkIndexWithSameSize() with a certain condition and then you don't need two methods